codegen/post_gen_hooks.go

```diff
@@ -21,116 +21,295 @@
+		fixtureMap := make(map[string]*Fixture, mockCount)
+		pathSymbolMap := make(map[string]string)
+		for _, instance := range clientInstances {
+			key := instance.ClassType + instance.InstanceName
+			var mc moduleConfig
+			if err := json.Unmarshal(instance.JSONFileRaw, &mc); err != nil {
 				return errors.Wrapf(
 					err,
-					"error generating mocks for client %q",
+					"error parsing client-config.json for client %q",
 					instance.InstanceName,
 				)
 			}
-			fmt.Printf(
-				"Generating %12s %12s %-30s in %-50s %d/%d\n",
-				"mock",
-				instance.ClassName,
-				instance.InstanceName,
-				path.Join(path.Base(h.CodeGenTargetPath()), instance.Directory, "mock-client"),
-				i+1, mockCount,
+
+			importPath := instance.PackageInfo.GeneratedPackagePath
+			if instance.ClassType == custom {
+				importPath = mc.Config.CustomImportPath
+			}
+
+			importPathMap[key] = importPath
+
+			// gather all modules that need to generate fixture types
+			f := mc.Config.Fixture
+			if f != nil && f.Scenarios != nil {
+				pathSymbolMap[importPath] = clientInterface
+				fixtureMap[key] = f
+			}
+		}
+
+		// only run reflect program once to gather interface info for all clients
+		pkgs, err := ReflectInterface(bin.projRoot, pathSymbolMap)
+		if err != nil {
+			return errors.Wrap(err, "error parsing Client interfaces")
+		}
+
+		var idx int32 = 1
+		var files sync.Map
+		var wg sync.WaitGroup
+		ic := make(chan *ModuleInstance)
+		ec := make(chan error, mockCount)
+
+		// cap the num of goroutines to num of CPU, because
+		// bin.GenMock in each goroutine starts a sub process
+		n := runtime.GOMAXPROCS(0)
+		if n > mockCount {
+			n = mockCount
+		}
+		wg.Add(n)
+
+		for i := 0; i < n; i++ {
+			go func() {
+				defer wg.Done()
+
+				for {
+					instance, ok := <-ic
+					if !ok {
+						return
+					}
+
+					key := instance.ClassType + instance.InstanceName
+					buildDir := h.CodeGenTargetPath()
+					genDir := filepath.Join(buildDir, instance.Directory, "mock-client")
+
+					importPath := importPathMap[key]
+
+					// generate mock client, this starts a sub process
+					mock, err := bin.GenMock(importPath, "clientmock", clientInterface)
+					if err != nil {
+						ec <- errors.Wrapf(
+							err,
+							"error generating mocks for client %q",
+							instance.InstanceName,
+						)
+						return
+					}
+					files.Store(filepath.Join(genDir, "mock_client.go"), mock)
+
+					// generate fixture types and augmented mock client
+					if f, ok := fixtureMap[key]; ok {
+						types, augMock, err := bin.AugmentMockWithFixture(pkgs[importPath], f, clientInterface)
+						if err != nil {
+							ec <- errors.Wrapf(
+								err,
+								"error generating fixture types for client %q",
+								instance.InstanceName,
+							)
+							return
+						}
+
+						files.Store(filepath.Join(genDir, "types.go"), types)
+						files.Store(filepath.Join(genDir, "mock_client_with_fixture.go"), augMock)
+					}
+
+					printGenLine(
```

Comment:
nit: Is fmt.Printf thread safe? If it's a wrapper around C `printf` then the answer is probably no, in which case you'll get mixed output from threads sometimes.